### ISSUE TITLE
Multiple cursors (again)

### ISSUE THREAD

**Issue by traverseda** [neovim/neovim#7257]
See also  #211, #3845, #299

Multiple cursors. Even though other ways of doing things may be more powerful, there are still users who'd prefer to use them.

There are some definite problems with "virtual cursors" like in [vim-multiple-cursors](https://github.com/terryma/vim-multiple-cursors).

>Is not that easy. It also requires multiple buffers, like when you copy using (i.e.) six carets and then pastes accordingly in the six current positions.

>This also means every caret can have it's own selection and context, so the solution of "virtual carets" is unfeasible because when (for example) a plugin relies on a caret-position-change-event it will break thanks to the wrongly assumption that the event did indeed happen.

~ @Ivanca

There's already been plenty of discussion on this, so let's try to keep away from "why" (because people want it, and it's useful for some things) and focus on the "how". That is, how to expand vims underlying metaphors to support multiple cursors cleanly and powerfully.

**Comment by justinmk** [comment]
FWIW, I've been working on this.

**Comment by KillTheMule** [comment]
As an alternative, might I suggest `inccomand=nosplit`? I've seen people judging that a viable alternative.

**Comment by traverseda** [comment]
Awesome, glad to hear this is being worked on.

**Comment by wsdjeg** [comment]
wow,I like this feature. but when multiple cursor is enable, how to deal with popupmenu?

**Comment by justinmk** [comment]
Is there more than one way?

**Comment by balta2ar** [comment]
I don't mean to pollute the thread, just putting here my old message as a reminder about typical problems with vim-multiple-cursors to make sure different parts of knowledge are linked together (https://github.com/extr0py/oni/issues/184):
> typical issues and user requests that already took place in the past. To name a few off the top of my head: mouse support, yank/paste behavior, completion support, interaction with other plugins (hooks to disable certain plugins when multiple cursors are active).

As for inccomand, it's great but after using it for a while, it felt somewhat incomplete. For examples, this: https://github.com/neovim/neovim/issues/5608

**Comment by KillTheMule** [comment]
> As for inccomand, it's great but after using it for a while, it felt somewhat incomplete. For examples, this: #5608

Good news then, I'm working on that as part of https://github.com/neovim/neovim/pull/7231. It already works, but kind of conflicts with the search highlighting sometimes, which will need to be solved later on.

**Comment by justinmk** [comment]
>  mouse support, 

@balta2ar What do you mean by "mouse support":

- placing a cursor with the mouse ?
- cascading mouse motions to _all_ cursors ?
- something else?

> completion support

The only thing that makes sense to me: let the "primary cursor" operate the menu, and whatever choice is made there is duplicated at the other cursors.

> interaction with other plugins (hooks to disable certain plugins when multiple cursors are active).

Like a `MulticursorPre` autocmd ?

**Comment by balta2ar** [comment]
> placing a cursor with the mouse

That's exactly what I meant. Perhaps also removing a cursor if an existing cursor was clicked. In VSCode though I remember there was a problem with conflicting keybindings (ctrl+click would follow the definition instead of placing a cursor). https://github.com/Microsoft/vscode/issues/3091 Not saying it's a problem in vim, just pointing out gotchas in other editors.

> cascading mouse motions to all cursors

Hm, never thought of that. Like what actions? Selection?

> The only thing that makes sense to me: let the "primary cursor" operate the menu, and whatever choice is made there is duplicated at the other cursors

That sounds reasonable, though the devil could be in the details. I basically wanted to point out that it's a very huge bummer that completion menu cannot be used with current implementation of https://github.com/terryma/vim-multiple-cursors, while other editors do that (like VSCode). To me, it's probably issue number one right now (along with general instability and unpredictable behavior -- if you press something beyond the plugin's prime, vim's gonna burst with errors corrupting text along the way).

> Like a MulticursorPre autocmd ?

Probably yes, what I meant is this: https://github.com/terryma/vim-multiple-cursors#interactions-with-other-plugins
But again, it depends on the core implementation details. The reason why such mechanism is needed in vim-multiple-cursors because the plugin is somewhat naive and it's actions are observed by other plugins such as deoplete, neocomplete or YouCompleteMe and excessive (sometimes harmful) work is performed as a result. Maybe there are ways to implement it transparently to all the plugins, or maybe there could be a reason to expose some API, I just don't know at the moment. Maybe plugin authors have ideas about that, e.g. @Shougo or @Valloric.

As for copy/paste, two scenarios are intuitive: 1) copy "something" using one cursor, create N cursors, paste that "something" into N locations 2) create N cursors, copy something, move cursors, paste. Each cursor should paste it's own data (a part of that implemented in https://github.com/terryma/vim-multiple-cursors/pull/110).

Another minor "nice-to-have"s are:
* search for the whole next word if in Normal mode
* search for exact text if in Visual mode (both are currently working as expected in the plugin and have an option to toggle)
* convert search results into cursors (in a file or in a selection).
* mappings for actions like FindNext, GoBack (undo FindNext), SkipCurrent.

Looking through https://github.com/terryma/vim-multiple-cursors/issues may give more ideas about what users expect from a multiplecursor feature and the current state of things.

**Comment by justinmk** [comment]
> Another minor "nice-to-have"s are

All the "search" behaviors can be trivially scripted by exposing some API like `cursors_add()`, `cursors_get()`.

**Comment by balta2ar** [comment]
One more thing to keep in kind: how will multiple cursors work with plugins that extend navigation/selection capabilities or operate with text objects, e.g.:
https://github.com/terryma/vim-expand-region
https://github.com/kana/vim-textobj-user
https://github.com/wellle/targets.vim

**Comment by balta2ar** [comment]
Hi, @justinmk! I wonder how things are going with that PoC that you mentioned? I remember someone posting links to video recordings of a vim meetup in Berlin where you talked about your prototype and demonstrated it. That's great news indeed! Do you have any estimates or plans when it could go public? Maybe it could attract more contributors, maybe even students could offer their help as it was with the initial implementation of `inccommand` feature?

**Comment by justinmk** [comment]
Soon(ish).

**Comment by andreyorst** [comment]
I've read #211, #3845 and just would like to add some note about multicursor usage. It can bring native snippets support into neovim. If we will be able to add multi cursors with macroses or mappings and abbreviations to certain positions from normal mode, then basically it will become a snippet engine. I'm currently using my own silly snippet engine based on abbreviations, and hideously complex macroses, and with multi cursor support I will be able to do it much easier. Take a look:

Class snippet and getter+setter snippets:
![class](https://user-images.githubusercontent.com/19470159/38161104-1f8ef1c0-34d1-11e8-9ef6-3f4d6756b768.gif)

Main and for snippets:
![main_for](https://user-images.githubusercontent.com/19470159/38161066-c1d26f6c-34d0-11e8-8dcc-52c8aa5fd9d3.gif)

And this is how getter/setter is done:
![image](https://user-images.githubusercontent.com/19470159/38161099-15fb22fa-34d1-11e8-9af1-8f744240e57f.png)
And `for` snippet with jump markers:
![image](https://user-images.githubusercontent.com/19470159/38161123-8765913c-34d1-11e8-8633-c39eceb65712.png)
Abbreviation that triggers `for` apperiance and then triggers jump mapping:
![image](https://user-images.githubusercontent.com/19470159/38161204-09f511c6-34d3-11e8-821f-eb33bec9a3a5.png)
And this is how I jump:
![image](https://user-images.githubusercontent.com/19470159/38161129-a5995558-34d1-11e8-98af-d49ee35a7282.png)
_(—  Andrey you could just use Ultisnips ot snipmate or even visual studio code with neovim plugin!_
_—  No! Thats boring!)_

So I developed this a while ago, and it is usable, except that i need 2 different mappings to replace single placeholder and another to replace several placeholders simultaniously, but in VS Code (I did not accidentally mention it above) it is done with multicursor, like so:
![image](https://user-images.githubusercontent.com/19470159/38161189-bc4e14d6-34d2-11e8-9a29-83a6260a95c9.png)
and we can jump between them. And as it is multicrusor, it will change text everywhere, so I won't need to use abbreviation hacks like double `/` symbol that is being sed right before cursor `|` in `%s/\<placeholder\>|/g` to achieve the same behaviour.

I know it is complex tricks if you didn't done this before, and I believe that advanced multicursor usage will make such cool stuff easier to make, and I just wanted to mention it here, because just voting for feature is not enough for me.

**Comment by mg979** [comment]
What's the status on this? I'm working on a multiple cursors plugin myself and I'm pretty excited at the progress I'm making. For now it can't edit anything but still :).

Some random pics:

---------
Place cursors at words:

![Imgur](https://i.imgur.com/swo5zHd.gif)

-------
Select inside/around brackets/quotes/etc:

![Imgur](https://i.imgur.com/Nn9d0NS.gif)

-------
Alternate cursor/extend mode, motions (even %), reverse direction (as in visual mode) and extend from the back. At any time you can switch from extend to cursor mode and viceversa.

![Imgur](https://i.imgur.com/MWR9DqL.gif)

-------
Indipendent anchoring

![Imgur](https://i.imgur.com/kN3BYjU.gif)

------
Find words under cursor, add new words (patterns stack), navigate regions, skip them, add regions with regex(works with cursors too).

![Imgur](https://i.imgur.com/8MmJbD0.gif)

-------
Select all

![Imgur](https://i.imgur.com/7kPW9Vi.gif)

Still a work in progress but I'm excited. What do you say? Worth pursuing or progresses in this area have been made, that I'm not aware of?

**Comment by balta2ar** [comment]
@mg979 Writing such plugin is no doubt a great experience and a learning opportunity, and I love seeing your enthusiasm, but if you want to be pragmatic, I think contributing to an already existing and a very popular plugin https://github.com/terryma/vim-multiple-cursors/ could be more beneficial for the whole vim community. Have you considered that? I'm sure your skills will very valuable there.

**Comment by andreyorst** [comment]
@mg979 IIUYC the last gif shows mappings needed to use your plugin. Theese are hell lot of mappings, that may conflict with other plugins. I don't know how it will be implemented in neovim, but I believe that it shold be sort of separate multi cursor mode, like normal, insert, mc_normal, mc_visual, etc. It will be more generic and will fit better in vim filosophy.

But having a plugin that may extend this modes (if it will be implemented lise so) will be good point. Something like how surround.vim adds new semantics to exisiting modes, but for multicursor mode with multicursor features.

**Comment by mg979** [comment]
@balta2ar if I was satisfied with [vim-multiple-cursors](https://github.com/terryma/vim-multiple-cursors/) I wouldn't have started a new plugin. That plugin is limited in what it can do (it implements only a subset of the things you can do in other editors with multiple cursors), it's also slow and it doesn't work well with neovim in general(in my own experience). I'm sure vim can do much better than this.

@andreyorst it's true that mappings could be an issue. I'm mostly using \<buffer> mappings (if one wanted, only one non-\<buffer> mapping would be necessary, to start the mode), that are mapped when the visual-multi mode starts, and are unmapped when it is exited. Still, they could overwrite other \<buffer> mappings and clear them on exit. It's already possible to toggle motion mappings ('w', 'b', etc) and one would use the arrow keys to move cursors/extending selections, and normal keys for the rest (à la Sublime Text, that is). No problems there.

**Comment by wsdjeg** [comment]
@mg979 does this plugin host on github? 

vim-multiple-cursor sometimes does not work for me. so I also implement a new plugin in [spacevim](https://github.com/SpaceVim/SpaceVim/).

![](https://user-images.githubusercontent.com/13142418/37873892-c76afb14-2fea-11e8-9149-aa955b8265f1.gif)

iedit support two mode iedit-normal/iedit-insert

**Comment by mg979** [comment]
@wsdjeg I put it on github [here](https://github.com/mg979/vim-visual-multi), but it can't do much for now. There are no editing functions, and bugs.

But it can do this for now.
![Imgur](https://i.imgur.com/3IsZzF3.gif)

**Comment by balta2ar** [comment]
@wsdjeg I'd like to try iedit but I don't use SpaceVim. Is there a separate repo for iedit? Or what parts do I need to take out of SpaceVim into my configuration?

**Comment by andreyorst** [comment]
> Soon(ish).

@justinmk _when(ish)_  ';^)

Actually I've discovered [this](https://github.com/mawww/kakoune) and it blew my mind __*__
These multi-cursors are maybe the best implementation I've ever seen. Please say that we will be able to do something similar to this.

# 

__*__ By "blew my mind" I've actually meant that this project is really incredible, and exists already for like 7 years, and I've **never** heard of it anywhere, but this is what any Vim/NeoVim user must try, it is pure amazing.

**Comment by balta2ar** [comment]
What's also amazing is the plugin (https://github.com/mg979/vim-visual-multi) that @mg979 advertised earlier in this thread. I was a bit sceptical initially because there was no editing support back in April, but it was implemented and in practice it has exceeded my expectations. If you're using classical https://github.com/terryma/vim-multiple-cursors/ and you're not entirely satisfied with it, you got to try https://github.com/mg979/vim-visual-multi, it's a substantial step forward, thanks to @mg979.

**Comment by kurkale6ka** [comment]
I agree. with @andreyorst. kakoune is amazing. It has the best multi cursor implementation out there.

**Comment by tototest99** [comment]
Sorry for the noise, but thought it may be an interesting reference/reminder, other than for myself, for possible future contributions to Neovim:
[easing operation on selections with Ex commands](https://github.com/martanne/vis/wiki/Differences-from-Kakoune)
especially union/inter/add/sub…

**Comment by zlianon** [comment]
Any progress?

**Comment by maisiliym** [comment]
> Soon(ish).

Almost 3 years have gone by! That blew by quickly!
Li would like to be very grateful to see it in its current state, just to see which parts of the code @justinmk touched.
Just pushing a branch and mentionning it here would do, if a PR is seen as inappropriate.

**Comment by sambarluc** [comment]
I just wanted to add a comment, mostly to keep this thread alive.

I've been recently switching back to (neo)vim after a few years using sublime text and vs code. While I appreciate many things of neovim (I wouldn't be switching back to it otherwise!), I find that the lack of first-class support for multi-cursor selections and operations is a major drawback. In fact, it's hard to believe for me that there's still debate on whether it's a good idea or not, superior or not to regex or whatever, or that one should first master vi techniques. I think that multi-cursor operations are great, the reasons for that are well explained by the authors of kakoune and, probably more important, the success of such feature across different editors speaks for itself. Of course, they don't replace regex or other, just as a car doesn't replace a bike, nor a truck.

It's quite clear that the there's not much interest for such feature from the main developers of neovim, but I would strongly encourage them to try it out!

**Comment by balta2ar** [comment]
@sambarluc did you read this thread?  Try this plugin, suggested above: https://github.com/mg979/vim-visual-multi

I'm pretty sure it'll cover all your needs.

**Comment by sambarluc** [comment]
> @sambarluc did you read this thread? Try this plugin, suggested above: https://github.com/mg979/vim-visual-multi
> 
> I'm pretty sure it'll cover all your needs.

Yes, I read it, I'm using it. It's great, but it's not the same thing as having multi-cursors as part of the editor.

**Comment by cseelus** [comment]
@balta2ar I use that too. Still, its just a plugin. As @sambarluc stated, first class support for multiple cursors would be highly appreciated by many NeoVim users and would enable a lot of new use cases (see Kakoune for example).

**Comment by pianocomposer321** [comment]
@justinmk Is there any update on this? This thread has been open for 4 years now, and we still don’t have any indication of any progress being made. Now that 0.5 has been released, it seems like maybe this feature could be worked on?

(And please, don’t give me the whole “He’s not getting payed to do this, just be thankful that the project exists at al, blah blah blah” thing. I know he’s not being payed, but it would still be nice to at least know whether the feature is still planned. I’m not demanding that it be added, I’m asking for an update about whether it will be.)

**Comment by justinmk** [comment]
@pianocomposer321 Still planned. No timeline.

**Comment by LamprosPitsillos** [comment]
Lets hope 0.7 will bring us Multi cursors 😁

**Comment by gpanders** [comment]
Kitty proposed a new protocol to show multiple cursors at the terminal level, which means Neovim wouldn't need to create "virtual" cursors (on terminals that support this anyway) https://github.com/kovidgoyal/kitty/issues/8927

**Comment by jceb** [comment]
1000x merci 🥳
